api/api-cache.php

```diff
@@ -0,0 +1,132 @@
+<?php
+// these functions use $db, so scripts referencing them need to define it
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|null The first entry that matches or null if no entry exists
+ */
+function getFromApiCacheTable(string $endpoint)
+{
+    global $db;
+    $q = $db->query('SELECT * FROM api_cache WHERE endpoint = "' . $db->real_escape_string($endpoint) . '"');
+    if ($q) {
+        $r = $q->fetch_array(MYSQLI_ASSOC);
+        return $r;
+    }
+    return null;
+}
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint. Will not return if the entry is expired.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|false|null The first entry that matches, false if the entry is expired, or null if no entry exists
+ */
+function getCached(string $endpoint)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    if ($r) {
+        $config = $db->query('SELECT * FROM outgoing_request_cache_config WHERE id = ' . $r['config_id'])->fetch_array(MYSQLI_ASSOC);
+        $ttl = $config['ttl'];
+        $now = time();
+        $retrievalTime = strtotime($r['last_successfully_retrieved'] . 'UTC');
+        if ($now - $retrievalTime <= $ttl) {
+            return $r;
+        }
+        return false;
+    }
+    return null;
+}
+
+/**
+ * Adds or updates an entry into the cache table.
+ * @param string $endpoint The endpoint for the API request
+ * @param string $content The content of API request's response
+ * @param int|string $config The ID of the config or its name in the config table
+ */
+function insertCached(string $endpoint, string $content, $config)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    $queryString = "";
+    if ($r) {
+        $id = $r['id'];
+        $queryString = 'UPDATE api_cache 
+                        SET last_successfully_retrieved = UTC_TIMESTAMP, 
+                            last_attempted_retrieval = UTC_TIMESTAMP, 
+                            retry_count = 0, 
+                            content =\'' . $db->real_escape_string($content) . '\'
+                        WHERE id = ' . $id;
+
+    } else {
+        $configId = 1;
```

Comment:
We should _only_ ever allow insertion by id in this case (due to the way we use the cache, it's always going to be hardcoded, we can always just fetch the id, we want to prioritise consistency and security over developer time), so no need for a sentinal value or dynamic type checking, always require the method is called with a numeric id.

Reply:
Removed in 7cb4d48